File: cpp/113.cpp
#include <vector>
#include <iostream>
using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

/* 16 ms 69.96%	
 * 20 MB 50.00%
 */
class Solution {
public:
    vector<vector<int>> res;
    int sum;
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<int> stack;
        this->sum = sum;
        order(root, stack, 0);
        return res;
    }
    void order(TreeNode* root, vector<int>& stack,int sum) {
        if (root != nullptr) {
            stack.push_back(root->val);
            sum += root->val;
            if (root->left || root->right) {  // 是否是根节点
                order(root->left, stack, sum);
                order(root->right, stack, sum);
            } else {
                if (sum == this->sum) {
                    this->res.push_back(stack);  // push_back为拷贝操作
                }
            }
            stack.pop_back();
        }
    }
};

File: python/113.py
class Solution:        
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        self.res = []
        self.target = sum
        self.order(root, [], 0)
        return self.res
    
    def order(self, root, ls, sum):
        """
        52 ms	
        17.9 MB	
        """
        if root is not None:
            ls += [root.val]
            sum += root.val
            if root.left != None or root.right != None:  # 叶子节点判断
                self.inorder(root.left, ls[:], sum)  # ls[:]很浪费空间和时间, 因此order2的时候改成复用ls,加一个pop操作即可
                self.inorder(root.right, ls[:], sum)
            else:
                if sum == self.target:
                    # 这里会重复append
                    self.res.append(ls)

    def order(self, root, ls, sum):
        """
        44 ms	 72.95% 
        13.8 MB  100%
        """
        if root is not None:
            ls += [root.val]
            sum += root.val
            if root.left != None or root.right != None: 
                self.inorder(root.left, ls, sum)  # 此处做了修改
                self.inorder(root.right, ls, sum)
            else:
                if sum == self.target:
                    # 这里会重复append
                    self.res.append(ls[:])  # 这里还是copy操作,因为ls是引用
            ls.pop() # 弹出栈
            

    